torch_xla/core/xla_model.py

```diff
@@ -376,14 +457,73 @@ def _fetch_gradients(optimizer):
 
 
 def _get_all_reduce_token():
-  token = getattr(_TLS, 'all_reduce_token', None)
+  devctx = _get_device_context()
+  token = getattr(devctx, 'all_reduce_token', None)
   if token is None:
     token = torch_xla._XLAC._xla_create_token()
-    _TLS.all_reduce_token = token
-  return token
+    devctx.all_reduce_token = token
+  return token, devctx
+
+
+def _torch_all_reduce(reduce_type, inputs, group=None):
+  import torch.distributed as dist
+
+  if reduce_type == REDUCE_SUM:
+    reduce_op = dist.ReduceOp.SUM
+  elif reduce_type == REDUCE_MUL:
+    reduce_op = dist.ReduceOp.PRODUCT
+  elif reduce_type == REDUCE_MIN:
+    reduce_op = dist.ReduceOp.MIN
+  elif reduce_type == REDUCE_MAX:
+    reduce_op = dist.ReduceOp.MAX
+  elif reduce_type == REDUCE_OR:
+    reduce_op = dist.ReduceOp.BOR
+  elif reduce_type == REDUCE_AND:
+    reduce_op = dist.ReduceOp.BAND
+  else:
+    raise RuntimeError('Invalid reduce type: {}'.format(reduce_type))
+
+  results = []
+  async_op = None
+  for tensor in inputs:
+    # Use async flag to overlap pytorch reduce ops with XLA tensor fetches.
+    cpu_tensor = torch_xla._XLAC._xla_get_cpu_tensors([tensor])[0]
+    results.append(cpu_tensor)
+    if async_op is not None:
+      async_op.wait()
+    async_op = dist.all_reduce(
+        cpu_tensor, reduce_op, async_op=True, group=group)
+  if async_op is not None:
+    async_op.wait()
+  return results
+
+
+def _host_all_reduce(reduce_type, inputs, cctx, scale=None):
+  # Barrier must happen on all devices.
+  torch_xla._XLAC._xla_sync_multi(
+      inputs, devices=[], wait=True, sync_xla_data=True)
+
+  # Here we use the torch.distributed reductions only on one device in the
+  # replication set, and then use in graph fast interconnect reduction to
+  # transfer the result to all replication devices.
+  # One core per fast interconnect replica group does the torch.distributed
+  # reduction and post the result, while the others post zeros.
+  if cctx.is_reduce_host:
+    results = _torch_all_reduce(reduce_type, inputs, group=cctx.interhost_group)
+    for i in range(0, len(inputs)):
+      inputs[i].copy_(results[i])
+      if scale is not None:
+        inputs[i].mul_(scale)
+  else:
+    for tensor in inputs:
+      tensor.zero_()
+  if cctx.requires_intercore_reduce:
+    token, devctx = _get_all_reduce_token()
+    devctx.all_reduce_token = torch_xla._XLAC._xla_all_reduce_inplace(
+        REDUCE_SUM, inputs, token, 1.0, [])
 
 
-def all_reduce(reduce_type, inputs, scale=1.0, groups=None):
+def all_reduce(reduce_type, inputs, scale=1.0, groups=None, cctx=None):
```

Comment:
Here keep the original function, and look how the other collective APIs do with:
```Python
token, devctx = _get_all_reduce_token()
```

Reply:
Done, running CI to check if all is OK.